planner/transformation_rules.go

```diff
@@ -0,0 +1,118 @@
+)
+
+// FromRangeTransformationRule rewrites the sequence from() |> range()
+type FromRangeTransformationRule struct{}
+
+// Pattern returns a pattern of the form:
+//     range
+//       |
+//     from
+func (rule FromRangeTransformationRule) Pattern() Pattern {
+	return &TreePattern{
+		RootType: ProcedureKind("RangeKind"),
+		Predecessors: []Pattern{
+			&LeafPattern{
+				RootType: ProcedureKind("FromKind"),
+			},
+		},
+	}
+}
+
+// Rewrite performs the logical-to-physical transformation:
+//     range
+//       |     ->     FromRange()
+//     from
+func (rule FromRangeTransformationRule) Rewrite(node PlanNode) (PlanNode, bool) {
+	rangeNode := node.ProcedureSpec().(*RangeProcedureSpec)
+	fromNode := node.Predecessors()[0].ProcedureSpec().(*FromProcedureSpec)
+	return &PhysicalPlanNode{
+		Spec: &FromRangeProcedureSpec{
+			Bucket:   fromNode.Bucket,
+			BucketID: fromNode.BucketID,
+			Bounds:   rangeNode.Bounds,
+			TimeCol:  rangeNode.TimeCol,
+			StartCol: rangeNode.StartCol,
+			StopCol:  rangeNode.StopCol,
+		},
+	}, true
+}
+
+// FromTagFilterTransformationRule rewrites the sequence from() |> range() |> filter()
+type FromTagFilterTransformationRule struct {
+	visitor *PredicateVisitor
+	catalog *Catalog
+}
+
+// Mock catalog for testing purposes
+type Catalog struct {
+	isTag map[string]bool
+}
+
+func (c *Catalog) IsTag(name string) bool {
+	return c.isTag[name]
+}
+
+// PredicateVisitor is used to visit a filter's predicate, assert that it
+// is a conjuctive predicate, and store all referenced (tag, tag value) pairs.
+type PredicateVisitor struct {
+	conjuctive        bool
+	tagFilters        map[string]string
+	predicateFunction *semantic.FunctionExpression
+}
+
+func (v *PredicateVisitor) Visit(node semantic.Node) semantic.Visitor {
+	panic("Not Implemented")
+}
+
+func (v *PredicateVisitor) Done() {}
+
+// Pattern returns a pattern of the form:
+//     filter
+//       |
+//     from
+func (rule FromTagFilterTransformationRule) Pattern() Pattern {
+	return &TreePattern{
+		RootType: ProcedureKind("FilterKind"),
+		Predecessors: []Pattern{
+			&LeafPattern{
+				RootType: ProcedureKind("FromKind"),
+			},
+		},
+	}
+}
+
+// Rewrite performs the logical-to-physical transformation:
+//                 filter
+// ( tag0 == "something" AND <conjunctive predicate> )    ->    FromRangeTagFilter()
+//                   |
+//                 from
+func (rule FromTagFilterTransformationRule) Rewrite(node PlanNode) (PlanNode, bool) {
+	filterNode := node.ProcedureSpec().(*FilterProcedureSpec)
+	fromNode := node.Predecessors()[0].ProcedureSpec().(*FromProcedureSpec)
+
+	predicate := filterNode.Fn
+	semantic.Walk(rule.visitor, predicate)
+
+	if !rule.visitor.conjuctive || len(rule.visitor.tagFilters) == 0 {
+		return nil, false
+	}
+
+	tagFilters := make(map[string]string, len(rule.visitor.tagFilters))
+	for k, v := range rule.visitor.tagFilters {
+		tagFilters[k] = v
+	}
+
+	return &PhysicalPlanNode{
```

Comment:
So the rule doesn't set any of its predecessors or successors?  This doesn't seem right to me.  I think the rule needs to at least be able to set the successors of the new node it creates.

Reply:
So the way it is right now, the planner handles setting predecessors and successors instead of the rule itself. As I'm writing this I realize that this approach is very simple for rewriting leaf nodes. However non-leaf nodes aren't quite as simple. You may be right. This logic might have to live with the rule itself. We can discuss more.